java/templates/s2i/run

```diff
@@ -13,6 +13,45 @@ JAVA_OPTIONS="${JAVA_OPTIONS:+${JAVA_OPTIONS} }$(/opt/jolokia/jolokia-opts)"
 JVM_ARGS="${JVM_ARGS:+${JVM_ARGS} }${JAVA_OPTIONS}"
 export JAVA_OPTIONS JVM_ARGS
 
+# support JBoss Tools OpenShift debug environment variables
+export JAVA_DEBUG="${JAVA_DEBUG:-${DEBUG}}"
+if [ "x${JAVA_DEBUG}" != x ]; then
+  export JAVA_DEBUG_PORT="${JAVA_DEBUG_PORT:-${DEBUG_PORT}}"
+fi
+# SpringBoot also uses DEBUG variable, so shade it with SPRINGBOOT_DEBUG variable
+if [ "x${SPRINGBOOT_DEBUG}" != x ]; then
+	export DEBUG="${SPRINGBOOT_DEBUG}"
+else
+	unset DEBUG
+fi
+
+# check for SpringBoot exploded archive
+if [ -f "${DEPLOYMENTS_DIR}/META-INF/MANIFEST.MF" ]; then
+	echo "SpringBoot exploded archive detected..."
+	if [ "x${JAVA_DEBUG}" != x ]; then
+		# does archive container spring devtools in lib
+		devtools_present=`ls "${DEPLOYMENTS_DIR}"/BOOT-INF/lib/spring-boot-devtools-*.jar 2>/dev/null | wc -l | tr -d '[[:space:]]'`
+		if [ ${devtools_present} = 1 ]; then
+			echo "SpringBoot developer tools library detected..."
+			# setup classpath and start class as main class
+			if [ "x${JAVA_CLASSPATH}" = x ]; then
+				JAVA_CLASSPATH="${DEPLOYMENTS_DIR}/BOOT-INF/lib/*"
+			else
+				JAVA_CLASSPATH="${DEPLOYMENTS_DIR}/BOOT-INF/lib/*:${JAVA_CLASSPATH}"
+			fi
+			export JAVA_CLASSPATH="${DEPLOYMENTS_DIR}/BOOT-INF/classes/:${JAVA_CLASSPATH}"
+			# setup start class as main class
+			main_class=`sed -n 's/Start-Class: \(.*\)$/\1/p' "${DEPLOYMENTS_DIR}/META-INF/MANIFEST.MF" | tr -d '[[:space:]]'`
+		fi
+	fi
+	# default to SpringBoot's Main-Class
+	if [ "x${main_class}" = x ]; then
```

Comment:
The extraction makes it so that the same image can be run in debug mode by just setting the JAVA_DEBUG environment variable, and user could test non-debug execution by removing it.

SpringBoot docs make a reference to how CloudFoundry runs apps exploded, without explicitly mentioning why. I don't think we have to do the same, but left it like that in case I discovered any advantages later, which I haven't found yet. Maybe CloudFoundry does it's own hot deploy and restart or something along those lines?? 

I was thinking of minimizing work before launching the app. So, I am thinking the following should be done during assemble:

* devtools check to decide whether to explode the jar or not. Before digging into SB devtools and JarLauncher code I thought it could do reloads etc. without having to build a special SB jar with devtools inside, but clearly it's not possible. 
* doing these tasks in assemble/run keeps run-java.sh focused on launching Java apps, instead of introducing SpringBoot cruft in it, sort of a separation of concerns.